### PR TITLE
Do not insert an additional trackPageView if already done by twig function

### PR DESCRIPTION
**Pull request by MHundeloh**
Do not insert an additional trackPageView if already done by twig function piwik('trackPageView',...) e.g. for setting custom dimensions only at page view track

### CODE REVIEW

**Comment by MHundeloh** [comment]
Or do i have rearrange the paqs entries so that trackPageView will be the last in array?

**Review by sebastiankugler — CHANGES_REQUESTED** [review]
I think the change is technically OK. If $this->paqs already contains a trackPageView (that may have been added using the twig piwik function/piwikPush), we don't want to add a second one.

However, it would be good if we could separate the two cases better (don't add trackPageView for trackSiteSearch pages, don't add trackPageView again if added manually). The comment inside the if block is only about the trackSiteSearch situation, this may be confusing for future readers.

**Comment by MHundeloh** [comment]
I separated trackPageView handling from trackSiteSearch handling.

File: Twig/Extension.php [hunk 1]
```diff
@@ -107,8 +107,17 @@ public function piwikCode()
     private function addDefaultApiCalls()
     {
         $this->paqs[] = ['enableLinkTracking'];
-
-        foreach ($this->paqs as $paq) {
+        $lastPaq = [];
```
Comment (sebastiankugler) [inline Twig/Extension.php:110]: Could you rename this variable to make its intention clearer? e. g. $manuallyAddedTrackPageView

Comment (sebastiankugler) [reply Twig/Extension.php:110]: I would also suggest to initialize it as `null`

File: Twig/Extension.php [hunk 2]
```diff
@@ -118,7 +127,11 @@ private function addDefaultApiCalls()
                 return; // Do not add 'trackPageView'
             }
         }
-
-        $this->paqs[] = ['trackPageView'];
+        if (!empty($lastPaq)) {
```
Comment (sebastiankugler) [inline Twig/Extension.php:130]: If you initialize with `null`, you can change this check to `null !== $manuallyAddedTrackPageView`

File: Twig/Extension.php [hunk 3]
```diff
@@ -118,7 +127,11 @@ private function addDefaultApiCalls()
...
+            $this->paqs[] = $lastPaq; // move manually added 'trackPageView' to the end
+            $this->paqs = array_values($this->paqs);
```
Comment (sebastiankugler) [inline Twig/Extension.php:132]: (Why) do you need this?

**Review by sebastiankugler — CHANGES_REQUESTED** [review]
Thanks a lot for your changes! I found the code a bit difficult to understand and suggested some changes to improve readability.

I also just realized that now the function addDefaultApiCalls does more than the name suggests (it adds and rearranges the calls). However, I don't know if we can easily rename the method or if it is part of the API (it's not declared final)

File: Twig/Extension.php (hunk 3 above)
Comment (relthyg) [reply Twig/Extension.php:132]: `array_values()` indexes an array numerically. The keys in `$this->paqs` might not follow a consecutive order at this point, and reindexing the array with `array_values` fixes that. It's more of an eye-candy purpose in this situation.

**Review by relthyg — CHANGES_REQUESTED** [review]
Please

- find a more speaking name for the variable `$lastPaq`
- initialize it as null and not as an empty array (and check against null further along)
- remove the reindexing of `$this->paqs`

all as mentioned in seperate comments

**Comment by relthyg** [comment]
Is there a chance you will continue with this in the near future, @MHundeloh ?

**Pull request merged**